pkg/cmd/server/kubernetes/master_config.go

```diff
@@ -402,58 +502,123 @@ func BuildKubernetesMasterConfig(
+	if masterConfig.DNSConfig != nil {
+		_, dnsPortStr, err := net.SplitHostPort(masterConfig.DNSConfig.BindAddress)
 		if err != nil {
-			return nil, fmt.Errorf("unable to parse DNS bind address %s: %v", options.DNSConfig.BindAddress, err)
+			return nil, fmt.Errorf("unable to parse DNS bind address %s: %v", masterConfig.DNSConfig.BindAddress, err)
 		}
 		dnsPort, err := strconv.Atoi(dnsPortStr)
 		if err != nil {
 			return nil, fmt.Errorf("invalid DNS port: %v", err)
 		}
-		m.ExtraServicePorts = append(m.ExtraServicePorts,
+		kubeApiserverConfig.ExtraServicePorts = append(kubeApiserverConfig.ExtraServicePorts,
 			kapi.ServicePort{Name: "dns", Port: 53, Protocol: kapi.ProtocolUDP, TargetPort: intstr.FromInt(dnsPort)},
 			kapi.ServicePort{Name: "dns-tcp", Port: 53, Protocol: kapi.ProtocolTCP, TargetPort: intstr.FromInt(dnsPort)},
 		)
-		m.ExtraEndpointPorts = append(m.ExtraEndpointPorts,
+		kubeApiserverConfig.ExtraEndpointPorts = append(kubeApiserverConfig.ExtraEndpointPorts,
 			kapi.EndpointPort{Name: "dns", Port: int32(dnsPort), Protocol: kapi.ProtocolUDP},
 			kapi.EndpointPort{Name: "dns-tcp", Port: int32(dnsPort), Protocol: kapi.ProtocolTCP},
 		)
 	}
 
+	return kubeApiserverConfig, nil
+}
+
+// TODO this function's parameters need to be refactored
+func BuildKubernetesMasterConfig(
+	masterConfig configapi.MasterConfig,
+	requestContextMapper apirequest.RequestContextMapper,
+	kubeClient kclientset.Interface,
+	internalKubeClient kinternalclientset.Interface,
+	informers shared.InformerFactory,
+	admissionControl admission.Interface,
+	originAuthenticator authenticator.Request,
+	kubeAuthorizer authorizer.Authorizer,
+) (*MasterConfig, error) {
+	if masterConfig.KubernetesMasterConfig == nil {
+		return nil, errors.New("insufficient information to build KubernetesMasterConfig")
+	}
+
+	controllerServer, cloud, err := buildControllerManagerServer(masterConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	schedulerServer, err := buildSchedulerServer(masterConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	apiserverConfig, err := buildKubeApiserverConfig(
+		masterConfig,
+		requestContextMapper,
+		kubeClient,
+		internalKubeClient,
+		informers,
+		admissionControl,
+		originAuthenticator,
+		kubeAuthorizer)
+	if err != nil {
+		return nil, err
+	}
+
 	kmaster := &MasterConfig{
-		Options:    *options.KubernetesMasterConfig,
-		KubeClient: kubeClient,
+		Options:            *masterConfig.KubernetesMasterConfig,
+		InternalKubeClient: internalKubeClient,
+		KubeClient:         kubeClient,
 
-		Master:            m,
-		ControllerManager: cmserver,
+		Master:            apiserverConfig,
+		ControllerManager: controllerServer,
 		CloudProvider:     cloud,
-		SchedulerServer:   schedulerserver,
+		SchedulerServer:   schedulerServer,
 		Informers:         informers,
 	}
 
 	return kmaster, nil
 }
 
+func ClientCARegistrationHook(options *configapi.MasterConfig) (*master.ClientCARegistrationHook, error) {
```

Comment:
No, hardcoded in the master. As long as we do not decompose or just plainly use `master.BuildMasterConfig`, we have to reimplement it: https://github.com/sttts/origin/pull/70/commits/c12bd72a82c0bf60265050d02af3624560d28fff